x/gamm/pool-models/balancer/pool.go

```diff
@@ -662,6 +662,19 @@ func (p *Pool) JoinPool(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (n
 	return numShares, nil
 }
 
+// JoinPoolNoSwap calculates the number of shares needed for an all-asset join given tokensIn with swapFee applied.
+// It updates the liquidity if the pool is joined successfully. If not, returns error.
+func (p *Pool) JoinPoolNoSwap(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error) {
```

Comment:
I believe both are already tested for at the keeper lever – do you think we should add tests in addition to those for balancer specifically? 

I'm in support of this but probably worth taking it to a separate small issue/PR on adding tests for balancer implementations of pool interface methods. Will make this issue if you agree.

Reply:
I think we should have a direct test for every method or function, especially, if they are exported. This is helpful in catching bugs early.

It would be great to have tests for methods/functions once they are introduced to avoid having them missed

Reply:
I tried implementing this and it ended up becoming a massive chunk of copy-pasted tests since both `JoinPool` and `JoinPoolNoSwap` are pretty much identical with `CalcJoinPool` and `CalcJoinPoolNoSwap`. Since this is a broader topic outside of the scope of this PR, I made a new issue and will add those changes to a new PR there.

Issue: https://github.com/osmosis-labs/osmosis/issues/2343

Reply:
I'm okay with tackling this in a separate PR. However, do you mind tackling that before v12, please? I already added it to the milestone.

I know it looks good visually but I would really prefer having maximum direct test coverage in gamm for v12

Reply:
Agreed, will assign myself to it right now. Thanks for adding it to the V12 blocker list